doc/developer/design/20251015_builtin_schema_migration.md

```diff
@@ -0,0 +1,152 @@
+
+- Builtin schema migration, when performed in read-only mode, does not modify durable state in a way that prevents processes at other versions from starting up.
+  "Other versions" here explicitly includes older versions.
+- Builtin schema migration, when performed in leader mode, does not interfere with schema migration performed by read-only processes with higher versions.
+  It may fence out processes running at lower versions.
+- Each Materialize version explicitly defines in its code the builtin schema migrations to be performed to upgrade existing durable state to this version.
+  If a builtin storage collection received a schema change without a corresponding migration instruction, upgrading to this version fails with an error.
+
+Additionally, in anticipation of the use-case-isolation work:
+
+- Builtin schema migration supports multiple concurrent processes executing the migration mechanism at the same time.
+
+## Out of Scope
+
+- Data migration of builtin collections, i.e. arbitrary rewriting of existing contents.
+
+## Solution Proposal
+
+The existing migration mechanism supports two approaches to schema migration:
+  - For backward-compatible schema changes (i.e. data written with the old schema remains readable with the new schema) it uses persist schema evolution.
+  - For incompatible schema changes it creates a new persist shard, effectively truncating the builtin storage collection.
+
+Both approaches have their merits, the first allowing schema changes without data loss, the second allowing breaking schema changes (at the cost of data loss).
+They are, however, problematic in how they are implemented:
+  - Schema evolution durably registers the new schema with the persist shard, globally changing the shard's schema and thus potentially prevent older versions from performing migration by schema evolution.
+  - Shard replacement uses a shared migration shard whose semantics are such that lower versions get fenced out by the existence of entries written by higher versions.
+
+The proposed solution keeps both migration approaches but avoids the problematic implementation details.
+
+### Schema Evolution
+
+A read-only process performing builtin schema migration by schema evolution skips registering the new schema with the persist shard.
+Instead, it only performs a sanity check ensuring that the new schema is indeed backward compatible with the current shard schema.
+Doing so requires no writes to durable state, and therefore doesn't interfere with other processes in any way.
+
+In the subsequent read-only bootstrap phase, the process creates persist read and write handles using the new schema.
+Read handles perform transparent migration of any data updates that flow through them, so dataflow hydration can proceed using the new schema.
+Write handles only require a matching registered shard schema when writing batches, which is something a read-only process doesn't do.
```

Comment:
Two qualifiers:
- Right now we require a matching registered schema when creating the handle, though as discussed that can change.
- I seem to recall a case in txn-wal where we write empty batches to advance the frontier, and we'd need to make sure that those use the old schema or work differently. (That seemed sketchy to me at the time, so I feel good about changing it.)

Reply:
Yep, I was pretending that we've already made the write handle change to not distract the reader unnecessarily, optimistically assuming that it's easy to implement and will be done before the design doc merges.

> I seem to recall a case in txn-wal where we write empty batches to advance the frontier, and we'd need to make sure that those use the old schema or work differently.

Interesting! So far I've only found that for _replaced_ builtin table shards, we tick forward their frontiers. But we explicitly make sure that they don't get inserted into txn-wal while the environment is in read-only mode.

Reply:
The relevant method is [`DataSnapshot::unblock_read`](https://github.com/MaterializeInc/materialize/blob/e757b4d11b49e7abf78ec533d1a90aa8c25c04a2/src/txn-wal/src/txn_read.rs#L65) - it's documented to do a CaA in some cases, if you trace the callers, you can see that quite a few methods which "look" read-only (snapshot-and-fetch, etc.) do end up calling it.

My memory is that those read-only-looking methods were sometimes called in read-only mode, but I haven't verified this anytime recently.

Personally I don't think unblock_read has to work that way. If this ends up being a real issue and not just a historical one we can probably discuss options then...